libgadget/cooling_uvfluc.c

```diff
@@ -177,6 +179,85 @@ struct UVBG get_local_UVBG(double redshift, double * Pos, const double * PosOffs
     return uvbg;
 }
 
+// TODO (jdavies): have some sort of flag to switch between UBVG models:
+// also, redshift and PosOffset argument not used yet
+static struct UVBG get_local_UVBG_from_J21(double redshift, double * Pos, const double * PosOffset, int heiiionized, double J21) {
+    /*int ind[3] = {-1};
+    for (int ii = 0; ii<3; ii++)
+    {
+        ind[ii] = pos_to_ngp(Pos[ii], PosOffset[ii], All.BoxSize, All.UVBGdim);
+    }*/
+
+    struct UVBG uvbg = {0};
+    
+    // N.B. J21 must be in units of 1e-21 erg s-1 Hz-1 (proper cm)-2 sr-1
+    uvbg.J_UV = J21;
+
+    //TODO(if local alpha desired, set J21 coeffs here with set_J21_coeffs(alpha)
+
+    //TODO:(jdavies) check if helium should be ionised here (once/twice)
+    //TODO:(jdavies) also check helium heating because there is a special case in the code
+    
+    //interpolators in cooling_rates.c should now be rate coeffs
+    //it seems a bit wasteful to calculate this for every particle
+    //but the global uv does an interpolation every time and this allows
+    //for future inhomogeneous alpha
+    struct J21_coeffs J21toUV = get_J21_coeffs(All.AlphaUV);
+
+    uvbg.gJH0   = J21toUV.gJH0 * J21; // s-1
+    uvbg.epsH0  = J21toUV.epsH0 * J21 * 1.60218e-12;  // erg s-1
+    uvbg.gJHe0  = J21toUV.gJHe0 * J21; // s-1
+    uvbg.epsHe0 = J21toUV.epsHe0 * J21 * 1.60218e-12;  // erg s-1
+
+    //TODO:look more into the qso_lightup model to see if this is correct
+    //TODO: set heating to be consistent with qso model, which activates long_mfp_heating
+    //a self-consistent approach probably needs separate stellar/quasar rates, H+ He+, He++ spectral slopes
+    if(heiiionized){
+        uvbg.gJHep  = J21toUV.gJHep * J21; // s-1
+        uvbg.epsHep = J21toUV.epsHep * J21 * 1.60218e-12;  // erg s-1
```

Comment:
Since this excursion set doesn't include quasars yet, I've set those rates to zero. I've also added to the conditional in get_local_UVBG to use the global rates for HeIII ionised particles. There may be some continuity issues if there is significant overlap between the end of hydrogen reionisation and the start of helium reionisation, but getting around this means I'd have to link the two models.